text/0000-run-to-completion-async-fn.md

```diff
@@ -0,0 +1,908 @@
+to `Future`s in the pre `async/await` era. `Future`s provide a `.poll()` method
+which can be used to drive `Future`s to completion and to query a `Future`s
+execution status. `Future`s can be dropped at any point in time to cancel an
+ongoing async operation.
+
+The latter ability made cancellation in async rust a lot easier to achieve than
+in most other environments. However the design choice also introduced some
+gaps, which the new async function type aims to fill.
+
+## Support for completion based operations
+
+Todays `Future`s do not allow us to support completion based operations in
+a flexible and safe fashion.
+
+**Completion based operations are operations which follow the follow sequence:**
+1. We start an asynchronous request, and provide references to all necessary
+  resources to the engine which executes the requests
+2. We later receive a notification from the engine that the request finished.
+  This notification can be delivered in a variety of ways: It could be callback
+  issued on a background thread, a signal or interrupt handler,
+  or it could be dequeued through a completion port/queue on an arbitrary thread.
+3. We acquire the result of the operation, which finalizes the operation.
+  after this step all resources which had been borrowed for the duration of the
+  async operation can be reused.
+
+**Example for such operations are:**
+- IO operations provided through the IO completion port (IOCP) APIs on windows.
+- IO operations provided through [`io-uring`](https://kernel.dk/io_uring.pdf)
+  and [`io_submit`](https://manpages.debian.org/testing/manpages-dev/io_submit.2.en.html) 
+  on Linux
+- IO operations offered through `libusb`
+- IO operations executed in Kernel space (e.g. disk command queues, etc)
+- Operations that are composed on top of the described promitives
+
+Async/await is theoretically powerful enough that it allows us to abstract over
+such APIs with code like:
+
+```rust
+/// Transmit a single byte array using an IO completion based low-level API
+async fn transmit_data() {
+    let buffer: String = "Data to transfer".into();
+    let data = buffer.as_bytes();
+    let bytes_transferred = engine.send(data).await;
+}
+```
+
+This is however not possible with todays `async fn`s. The reason for this is
+that the `Future` produced by invoking the function can be dropped while the
+execution is still in progress inside `engine.send(data).await`.
+If this happens, the engine (which still would hold onto a pointer to the data)
+would act on memory that does not represent the actual data to transmit anymore.
+This would lead to undefined behavior.
+
+One workaround to this problem is to use owned objects within completion based
+operations. E.g.
+
+```rust
+async fn transmit_data() {
+    let buffer: String = "Data to transfer".into();
+    let data: Bytes = buffer.into();
+    let (bytes_transferred, data) = engine.send(data).await;
```

Comment:
It mainly solves the problem of corrupting memory if the operation is aborted. With this workaround the memory is temporarily leaked while the operation continues running in the background, but the overall operation is safe.
As you mentioned, it might not solve other logic issues.